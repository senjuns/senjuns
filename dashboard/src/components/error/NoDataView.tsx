import { FC } from 'react';
import styled from 'styled-components';

import { ReactComponent as CalendarDropDownIcon } from '../../assets/svg/calendar-dropdown.svg';
import { ReactComponent as WarningIcon } from '../../assets/svg/warning.svg';
import { Typography } from '../../components/common';
import { CropCycleMenu } from '../../components/common/CropCycleMenu';
import { useCropCycleDetails } from '../../contexts/CropCycleProvider';
import { getCurrentDateTime } from '../../shared/utils';

const DEFAULT_MESSAGE =
  'There is no data available for this time range. Please, select a different one to continue.';

interface NoDataViewProps {
  message?: string;
}

const NoDataView: FC<NoDataViewProps> = ({ message = DEFAULT_MESSAGE }) => {
  const {
    startTime,
    range: cropCycleRange,
    onChange: onChangeCropCycleRange,
  } = useCropCycleDetails();

  return (
    <NoDataContainer>
      <View>
        <WarningIconContainer>
          <WarningIcon />
        </WarningIconContainer>

        <NoDataLabel variant="h5" fontWeight="bold">
          No data available
        </NoDataLabel>

        <Description color="gray" variant="body2">
          {message}
        </Description>

        {startTime && (
          <CropCycleMenu
            current={cropCycleRange}
            isInvalid={false}
            defaultLabel={<CalendarDropDownIcon />}
            whole={{
              start: startTime || getCurrentDateTime(),
              end: getCurrentDateTime(),
            }}
            onSelectRange={onChangeCropCycleRange}
          />
        )}
      </View>
    </NoDataContainer>
  );
};

export default NoDataView;

const NoDataContainer = styled.div`
  min-height: 300px;
  height: 100%;
  width: 100%;
  padding: 40px;
  background: #f8f8fc;
  box-sizing: border-box;
  border-radius: 40px;
  display: flex;
  justify-content: center;
`;

const View = styled.div`
  width: 320px;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;

const WarningIconContainer = styled.div`
  margin-bottom: 28px;
`;

const NoDataLabel = styled(Typography)`
  margin-bottom: 10px;
`;

const Description = styled(Typography)`
  margin-bottom: 20px;
  text-align: center;
`;
