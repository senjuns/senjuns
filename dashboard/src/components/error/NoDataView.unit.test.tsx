// eslint-disable-next-line import/no-extraneous-dependencies
import * as ShallowRenderer from 'react-test-renderer/shallow';
import NoDataView from './NoDataView';

jest.mock('../../shared/utils', () => ({
  getCurrentDateTime: () => new Date('2022-02-14T00:00:00'),
}));

jest.mock('../../contexts/CropCycleProvider', () => ({
  useCropCycleDetails: () => ({
    startTime: new Date(2022, 2, 12),
    range: { start: new Date(2022, 2, 1), end: new Date(2022, 2, 14) },
    onChange: jest.fn(),
  }),
}));

describe('<NoDataView />', () => {
  it('renders correctly.', async () => {
    const renderer = ShallowRenderer.createRenderer();
    renderer.render(<NoDataView />);

    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
});
