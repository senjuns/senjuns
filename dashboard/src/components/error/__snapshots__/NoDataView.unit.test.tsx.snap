// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NoDataView /> renders correctly. 1`] = `
<styled.div>
  <styled.div>
    <styled.div>
      <ForwardRef(SvgWarning) />
    </styled.div>
    <Styled(Typography)
      fontWeight="bold"
      variant="h5"
    >
      No data available
    </Styled(Typography)>
    <Styled(Typography)
      color="gray"
      variant="body2"
    >
      There is no data available for this time range. Please, select a different one to continue.
    </Styled(Typography)>
    <CropCycleMenu
      current={
        Object {
          "end": 2022-03-14T00:00:00.000Z,
          "start": 2022-03-01T00:00:00.000Z,
        }
      }
      defaultLabel={<ForwardRef(SvgCalendarDropdown) />}
      isInvalid={false}
      onSelectRange={[MockFunction]}
      whole={
        Object {
          "end": 2022-02-14T00:00:00.000Z,
          "start": 2022-03-12T00:00:00.000Z,
        }
      }
    />
  </styled.div>
</styled.div>
`;
